`T | None` narrowing not propagated into list comprehensions

### Description

After `if r is None { return ...; }`, `r` is correctly narrowed to `R` in the function body - direct attribute access works. But the same attribute access inside a list comprehension on the next line fails `E1099`. Other type checkers (pyright, mypy) propagate narrowing into comprehensions; Jac's pass does not.

## Repro

```jac
obj R {
    has parts: list[str] = [];
}

def use(r: R | None) -> list[str] {
    if r is None { return []; }
    direct: list[str] = r.parts;                                     # ✓ ok
    return [r.parts[i] for i in range(len(r.parts))];                # ✗ E1099
}

with entry {
    print(use(R(parts=["a", "b"])));
}
```

## Observed

```
$ jac check narrow_test.jac
✖ Error: error[E1099]: Cannot access attribute "parts" for type "R | NoneType"
  --> narrow_test.jac:8:45     (inside list comprehension)

$ jac run narrow_test.jac
['a', 'b']
```

Type check fails, runtime succeeds - the code is correct; only the type-narrowing pass treats the comprehension as outside the enclosing narrowed scope.

--- a/docs/docs/community/release_notes/unreleased/jaclang/5951.bugfix.md
+++ b/docs/docs/community/release_notes/unreleased/jaclang/5951.bugfix.md
@@ -1,0 +1,1 @@
+- **Fix: type narrowing now reaches inside comprehensions**: `T | None` narrowing established by an enclosing guard propagates into list/dict/set/generator comprehension iterators, conditionals, and output expressions, eliminating spurious `E1099` false positives.

--- a/jac/jaclang/compiler/type_system/type_evaluator.impl/type_evaluator.impl.jac
+++ b/jac/jaclang/compiler/type_system/type_evaluator.impl/type_evaluator.impl.jac
@@ -3558,11 +3558,16 @@ impl TypeEvaluator.postinit -> None {
 # Type narrowing support (CFG-based, flow-sensitive)
 # Pyright-style implementation with affected_symbols optimization.
 # -------------------------------------------------------------------------
-"""Find the nearest enclosing CFG node (CfgExpr or CodeBlockStmt)."""
+"""Find the nearest enclosing CFG node (CfgExpr or CodeBlockStmt).
+
+`InnerCompr` inherits `UniCFGNode` for its scope/symbol bookkeeping but is
+never wired into the CFG, so it has no `bb_in`/`bb_out`. Treat it as
+transparent so narrowing established by an enclosing guard reaches
+references inside the comprehension's iterator and conditional clauses."""
 def _find_cfg_node_for(at_node: uni.UniNode) -> uni.UniCFGNode | None {
     cur: uni.UniNode | None = at_node;
     while cur is not None {
-        if isinstance(cur, uni.UniCFGNode) {
+        if isinstance(cur, uni.UniCFGNode) and not isinstance(cur, uni.InnerCompr) {
             return cur;
         }
         cur = cur.parent;

--- a/jac/tests/compiler/passes/main/fixtures/checker/checker_bug_compr_narrowing.jac
+++ b/jac/tests/compiler/passes/main/fixtures/checker/checker_bug_compr_narrowing.jac
@@ -1,0 +1,57 @@
+"""Fixture: `T | None` narrowing must propagate into comprehension iterators.
+
+After `if r is None { return ...; }`, `r` is narrowed to `T` in the function body.
+Pyright and mypy propagate that narrowing into every sub-expression of a
+comprehension (out_expr, the iterator collection, and the conditional).
+Jac previously dropped the narrowing as soon as the lookup walked into an
+`InnerCompr` — an orphan UniCFGNode that is not wired into the CFG —
+producing E1099 false positives on attribute access inside the comprehension.
+
+After the fix, every attribute access below must type-check.
+"""
+
+obj R {
+    has parts: list[str] = [];
+}
+
+# Iterator collection: `r.parts` inside `range(len(...))`.
+def use_iter(r: R | None) -> list[str] {
+    if r is None { return []; }
+    return [str(i) for i in range(len(r.parts))];
+}
+
+# Out expression: `r.parts[i]` as the comprehension output.
+def use_out(r: R | None) -> list[str] {
+    if r is None { return []; }
+    return [r.parts[i] for i in range(5)];
+}
+
+# Conditional clause: narrowed `r` referenced in `if` filter.
+def use_cond(r: R | None) -> list[str] {
+    if r is None { return []; }
+    return [p for p in ["a", "b", "c"] if p in r.parts];
+}
+
+# Combined: all three positions in the same comprehension.
+def use_all(r: R | None) -> list[str] {
+    if r is None { return []; }
+    return [r.parts[i] for i in range(len(r.parts)) if r.parts[i] != ""];
+}
+
+# Dict comprehension: narrowing propagates into kv_pair, iterator, and condition.
+def use_dict(r: R | None) -> dict[int, str] {
+    if r is None { return {}; }
+    return {i: r.parts[i] for i in range(len(r.parts)) if r.parts[i] != ""};
+}
+
+# Set comprehension.
+def use_set(r: R | None) -> set[str] {
+    if r is None { return set(); }
+    return {r.parts[i] for i in range(len(r.parts))};
+}
+
+# Generator comprehension.
+def use_gen(r: R | None) -> list[str] {
+    if r is None { return []; }
+    return list(r.parts[i] for i in range(len(r.parts)));
+}

--- a/jac/tests/compiler/passes/main/test_checker_bugs.jac
+++ b/jac/tests/compiler/passes/main/test_checker_bugs.jac
@@ -654,3 +654,24 @@ test "bug_none_eq_neq_narrowing" {
         f"Got errors: {[str(e) for e in program.errors_had]}"
     );
 }
+
+# =========================================================================
+# Bug #COMPR-NARROW: `T | None` narrowing did not reach comprehension
+# iterators or conditionals.
+#
+# `InnerCompr` is a `UniCFGNode` but is never wired into the CFG. A
+# reference to a narrowed symbol inside the comprehension's `collection`
+# or `conditional` walks up to the orphan `InnerCompr` and stops, missing
+# the narrowing established by an enclosing guard. Pyright and mypy treat
+# the comprehension as part of the enclosing flow scope; Jac now does too.
+# =========================================================================
+test "bug_comprehension_narrowing" {
+    program = compile_and_check("checker_bug_compr_narrowing.jac");
+
+    assert len(program.errors_had) == 0 , (
+        "`T | None` narrowing established by an enclosing guard must "
+        "propagate into list/dict/set/gen comprehension iterators, out "
+        "expressions, and conditional clauses. "
+        f"Got errors: {[str(e) for e in program.errors_had]}"
+    );
+}

